src/jit/decomposelongs.cpp

```diff
@@ -532,25 +523,92 @@ void DecomposeLongs::DecomposeCast(GenTree** ppTree, Compiler::fgWalkData* data)
     GenTreeStmt* curStmt = m_compiler->compCurStmt->AsStmt();
 
     assert(tree->gtPrev == tree->gtGetOp1());
-    NYI_IF(tree->gtOverflow(), "TYP_LONG cast with overflow");
-    switch (tree->AsCast()->CastFromType())
+
+    var_types srcType = tree->AsCast()->CastFromType();
+    var_types dstType = tree->AsCast()->CastToType();
+
+    if (tree->gtFlags & GTF_UNSIGNED)
+    {
+        srcType = genUnsignedType(srcType);
+    }
+
+    if (varTypeIsLong(srcType))
     {
-    case TYP_INT:
-        if (tree->gtFlags & GTF_UNSIGNED)
+        if (tree->gtOverflow() && (varTypeIsUnsigned(srcType) != varTypeIsUnsigned(dstType)))
         {
-            loResult = tree->gtGetOp1();
-            hiResult = new (m_compiler, GT_CNS_INT) GenTreeIntCon(TYP_INT, 0);
-            m_compiler->fgSnipNode(curStmt, tree);
+            GenTree* srcOp = tree->gtGetOp1();
```

Comment:
Yes, it's kind of confusing and I'm not 100% sure it happens all the time. I suppose it's done like this because otherwise we'd end up with a dangling hi tree.